File: src/test/java/mk/jug/collectors/HelperCollectors.java
package mk.jug.collectors;

import mk.jug.collectors.model.Player;
import org.junit.Test;

import java.util.*;
import java.util.stream.Collectors;

import static org.hamcrest.CoreMatchers.containsString;
import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.MatcherAssert.assertThat;

/**
 * Some examples with {@link java.util.stream.Stream} and {@link Collectors}
 */
public class HelperCollectors extends BaseTest {

    @Test
    public void allPlayers() {
        print(players);
    }

    @Test
    public void collectToList() {
        List<Player> moreThan25K = players.stream()
                .filter(each -> each.getPoints() > 25_000)
                .collect(Collectors.toList());
        print(moreThan25K);

        assertThat(moreThan25K.size(), is(20));
    }


    @Test
    public void groupByKPoints() {
        Map<Integer, List<Player>> byKPoints = players.stream()
                .collect(Collectors.groupingBy(player -> player.getPoints() / 1_000));

        byKPoints.entrySet().forEach(PRINT_MAP);
        assertThat(byKPoints.size(), is(16));
    }

    @Test
    public void groupByKPointsOrdered() {
        Map<Integer, List<Player>> byKPoints = players.stream()
                .collect(Collectors.groupingBy(
                        player -> player.getPoints() / 1_000,
                        () -> new TreeMap<>(Comparator.<Integer>naturalOrder().reversed()),
                        Collectors.toList())
                );

        byKPoints.entrySet().forEach(PRINT_MAP);
        assertThat(byKPoints.size(), is(16));
    }

    @Test
    public void mostPoints() {
        Optional<Player> max = players.stream()
                .max(Comparator.comparing(Player::getPoints));
        max.ifPresent(System.out::println);

        assertThat("Exist", max.isPresent());
        max.ifPresent(player -> assertThat(player.getPoints(), is(38_387)));
    }

    @Test
    public void pointsStatistics() {
        IntSummaryStatistics pointsStatistics = players.stream()
                .collect(Collectors.summarizingInt(Player::getPoints));

        /*pointsStatistics = players.stream()
                .mapToInt(Player::getPoints)
                .summaryStatistics();*/

        System.out.println(pointsStatistics);
    }

    @Test
    public void bestByFirstLetter() {
        Comparator<Player> byPoints = Comparator.comparing(Player::getPoints);

        Map<Character, Optional<Player>> bestByFirstLetter = players.stream()
                .collect(
                        Collectors.groupingBy(
                                each -> each.getName().charAt(0),
                                //Collectors.reducing(BinaryOperator.maxBy(byPoints))
                                Collectors.maxBy(byPoints)
                        )
                );
        bestByFirstLetter.entrySet().forEach(each ->
                System.out.printf("%c -> %s\n", each.getKey(), each.getValue().orElse(null)));

    }

    @Test
    public void groupGOATContenderNames() {
        String goats = players.stream()
                .filter(each -> each.getPoints() > 30_000)
                .map(Player::getName)
                .collect(Collectors.joining(", ", "=> ", " <="));

        System.out.println(goats);
        assertThat(goats, containsString("Jordan"));
        assertThat(goats, containsString("Kobe"));
    }

}

File: src/test/kotlin/mk/jug/collectors/HelperCollectors.kt
package mk.jug.collectors

import mk.jug.collectors.model.Player
import mk.jug.collectors.model.SummaryStatisticsInt
import org.hamcrest.CoreMatchers.`is`
import org.hamcrest.CoreMatchers.containsString
import org.hamcrest.MatcherAssert.assertThat
import org.junit.Test
import java.util.*
import kotlin.comparisons.compareBy

/**
 * Collection examples using Kotlin Collections
 */
class HelperCollectors : BaseTest() {

    @Test
    fun allPlayers() {
        players.forEach(::println)
    }

    @Test
    fun collectToList() {
        val moreThan25K = players.filter { it -> it.points > 25000 }
        moreThan25K.forEach(::println)

        assertThat(moreThan25K.size, `is`(20))
    }

    @Test
    fun groupByKPoints() {
        val byKPoints = players.groupBy { it.points / 1000 }

        byKPoints.entries.forEach {
            println("=== ${it.key}K points ===")
            it.value.forEach(::println)
        }
        assertThat(byKPoints.size, `is`(16))
    }


    @Test
    fun groupByKPointsOrdered() {
        val byKPoints = players.groupByTo(LinkedHashMap<Int, MutableList<Player>>()
                .toSortedMap(compareBy <Int> { it }.reversed()))
        { it.points / 1000 }

        byKPoints.forEach {
            println("== ${it.key}K Points ==")
            it.value.forEach(::println)
        }
        assertThat(byKPoints.size, `is`(16))
    }

    @Test
    fun mostPoints() {
        val max = players.maxBy { it.points }
        assertThat("Exist", max != null)
        max?.let { assertThat(it.points, `is`(38387)) }
    }

    inline fun Collection<Int>.summarizingInt(): SummaryStatisticsInt =
            this.fold(SummaryStatisticsInt()) { stats, it -> stats.accumulate(it) }

    inline fun <T : Any> Collection<T>.summarizingInt(transform: (T) -> Int): SummaryStatisticsInt =
            this.fold(SummaryStatisticsInt()) { stats, it -> stats.accumulate(transform.invoke(it)) }

    @Test
    fun pointsStatistics() {
        val pointsStatistics = players.fold(SummaryStatisticsInt()) { stats, it -> stats.accumulate(it.points) }
        println(pointsStatistics)

        val stats = players.summarizingInt(Player::points)
        println(stats)
    }

    @Test
    fun bestByFirstLetter() {

        val bestByFirstLetter = playersSeq()
                .groupBy { it -> it.name[0] }
                .mapValues { list -> list.value.maxBy { it.points } }

        bestByFirstLetter.forEach { entry ->
            println("${entry.key} -> ${entry.value}")
        }

    }

    @Test
    fun groupGOATContenderNames() {
        val goats = playersSeq()
                .filter { it.points > 30000 }
                .map { it.name }
                .joinToString(", ", "=> ", " <=")

        println(goats)
        assertThat(goats, containsString("Jordan"))
        assertThat(goats, containsString("Kobe"))
    }

}
